Allow games to easily store things and store play time and scores

Joshix-1 opened the pull request:
This pr adds helper methods for easily storing things.

They are used for storing scores, play time and the time you last played. This data can be accessed by games with `GameManager.get_high_score()`, `GameManager.get_last_played()` and  `GameManager.get_played_time()`.

Games can easily store stuff too. They can just call `GameManager.get_game_data()` and then modify the returned dictionary. The changes will then be saved automatically saved when the game is ended.

To be future proof the pr also adds a `get_current_player()` method, which can be modified to return a player name. If this is done, all the storing functionality will still work.

example save data:
'~/.local/share/godot/app_userdata/Suffragium/savefiles/p/testgame/game_meta_data.json'
```json
{"last_played":1654971870,"played_time":42.159,"scores":[{"score":10,"_time":1654967221},{"score":7,"_time":1654967262}]}
```
'~/.local/share/godot/app_userdata/Suffragium/savefiles/p/testgame/game_data.json'
```json
{"num":1229874613}
```

ckanitz commented:
just curious: why did you prefix the meta keys with underscore?

Joshix-1 commented:
I don't know, I just had the feeling I would like that better

ASecondGuy commented:
I think the meta keys shouldn't have _ at the start because that indicates a "private" variable/function in Godot.

I think that savegame thing you are trying to do is a good idea. But it's rough around the edges.

What you put on the cfg as meta with "_game_name" is actually the folder name that doesn't necessarily match the actual game name in the cfg. The folder name is also redundant because you can infer it from the folder path. Not quite sure how to feel about the highscore in meta though.

ASecondGuy commented on game/menu/gamedisplay.gd:
```diff
@@ -34,5 +34,17 @@ func setup(game_cfg: ConfigFile):
+		text += "\n\nHighscore: " + str(high_score)
+
+	# update the high_score displayed in $VBoxContainer/RichTextLabel
+	if $VBoxContainer/RichTextLabel.bbcode_text != text:
```
If you use a Node more than once in a function you should save it in a local variable.
Because this function might run many times you should put it in a onready var.

Joshix-1 commented:
> I think the meta keys shouldn't have _ at the start because that indicates a "private" variable/function in Godot.

Yeah, makes sense.

> What you put on the cfg as meta with "_game_name" is actually the folder name that doesn't necessarily match the actual game name in the cfg.

Yes, but I feel like this is the better key. (it is guaranteed to be unique, game_names could have duplicates) Naming it folder_name would be better

> The folder name is also redundant because you can infer it from the folder path.

I could maybe remove the folder_path and just set the folder_name

> Not quite sure how to feel about the highscore in meta though.

It's the easiest solution I could think of. I don't know how I would use the get_high_score method in the other file

ASecondGuy commented:
> It's the easiest solution I could think of. I don't know how I would use the get_high_score method in the other file

That is what GameManager is for. It should have a get_highscore function. Then you'd get a games highscore by calling: `GameManager.get_highscore("my Game")`

The Game name in the cfg is just a display name I guess. A unique id wasn't needed before and I agree the folder name is the best option for that. But I also think we should implement naming conventions for the folders. Like, Your "Name-GameName". Then the games displayname can change whenever without breaking anything and it doesn't need to be unique.

Joshix-1 commented:
> That is what GameManager is for. It should have a get_highscore function. Then you'd get a games highscore by calling: GameManager.get_highscore("my Game")

thx, improved that (didn't know GameManager was available in this context, I'm still learning Godot)

> But I also think we should implement naming conventions for the folders.

Would be an idea, but I don't think that that is important, because we don't show the folder name anywhere.

> Then the games displayname can change whenever without breaking anything and it doesn't need to be unique.

That's currently the case. We don't need naming conventions for that. We should never change the folder names

ASecondGuy commented on game/menu/gamemanager.gd:
```diff
@@ -103,7 +109,7 @@ func _build_menu():
 	for game_name in GAMES.keys():
 		var display = _preview_scene.instance()
-		display.setup(GAMES[game_name])
+		display.setup(GAMES[game_name], self)
```
Instead of immediately doing setup and needing to hand over self you could use:
`display.call_deferred("setup", GAMES[game_name])`
This would delay the call until the idle frame starts when the child is already added so get_partent() will work.

Joshix-1 replied:
Why are you commenting on things I already changed?

ASecondGuy commented:
As an Autoload GameManager is always available everywhere. 
Except inside itself (duh) and sometimes in other Autoloads at startup (those that load before it).

Joshix-1 commented:
Now I'm using it too. It's really nice that you can access it everywhere. Then load_game_data and save_game_data can be used easily everywhere that is really nice.
What do you think now about the changes?

ASecondGuy commented:
I think its still a little bit confusing. 
Maybe scores should be completely handled by the GameManager.
As in. After end_game() receives the score it is saved no matter what the game does and there is no way for a game to save the scores themselves.

Joshix-1 commented:
> After end_game() receives the score it is saved no matter what the game does

That's the case

>  there is no way for a game to save the scores themselves.

How would that be possible? The game can access the GameManager...

The games should not use the load_game_data and save_game_data methods for scores. The scores are handled by the end_game method.

ASecondGuy commented:
But right now this is completely possible, is it not?
`GameManager.save_game_data("game_scores", "This_Game, some_score)`
This might even break something. (I haven't looked too far into this because I'm also writing documentation rn)

Joshix-1 commented:
Yes that's possible. And yes that breaks stuff. But I'm don't see a way to prevent that. The games can do everything the GameManager can do too.

ASecondGuy commented:
Of course. That's why you need to make it hard to screw up.
And If you can't, that is what the "private" functions are for. The stuff you shouldn't call unless you really know what you are doing.
But if you want people to have something but you can't make it easy enough to use your whole system might be flawed.

Joshix-1 commented:
> That's why you need to make it hard to screw up.
I think I'm doing that now.

>  But if you want people to have something but you can't make it easy enough to use your whole system might be flawed.

Could be. ¯\_(ツ)_/¯ 

I btw stole your documentation to document my additions (sorry lol)

Numenter commented on game/menu/gamedisplay.gd:
```diff
@@ -34,5 +39,27 @@ func setup(game_cfg: ConfigFile):
 	$InfoDialog/Container/descCont/Statslab.text = text
 
 
+func _on_VBoxContainer_draw():
```
When the values ​​only change on end_game(). May call this only on end_game() and not every draw for every game display.

Joshix-1 replied:
That's true and a great idea, ~~but how can I do that?~~

ASecondGuy reviewed (commented):
Please explain how the GAME_DATA_CACHE and all related functions work and what they are supposed to do. It is hard to see that.

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+## A Step by step Guide
+
+### 1. Chose a folder
+All files related only to your game go into a folder inside res://games/. You shouldn't rename it after the game is first merged. 
```
```suggestion
All files related only to your game go into a folder inside `res://games/`. You shouldn't rename it after the game is first merged. 
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+
+### 1. Chose a folder
+All files related only to your game go into a folder inside res://games/. You shouldn't rename it after the game is first merged. 
+There are no rules what to call it but a good Idea is to use your name and the games name (ex. "asecondguy_assimilator") to avoid any accidental collisions.
```
```suggestion
There are no rules what to call it but a good idea is to use your name and the games name (ex. "asecondguy_assimilator") to avoid any accidental collisions.
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+There are no rules what to call it but a good Idea is to use your name and the games name (ex. "asecondguy_assimilator") to avoid any accidental collisions.
+
+### 2. Make a game.cfg
+The game.cfg file resides in your Minigame's folder. It defines how your game is displayed. 
```
```suggestion
The game.cfg file resides in your minigame folder. It defines how your game is displayed. 
```

Joshix-1 replied:
are you sure? that's genitiv. I would say it is correct

BlobCodes replied:
1. you wrote it in uppercase
2. I'm pretty sure you wouldn't use the genitive in this case.
   Example: [in the godot docs you often see `project folder`](https://docs.godotengine.org/en/stable/classes/class_directory.html), not `project's folder`

Joshix-1 replied:
1. it wasn't me
2. makes sense

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+
+### 2. Make a game.cfg
+The game.cfg file resides in your Minigame's folder. It defines how your game is displayed. 
+If you are unsure copy the game.cfg from the testgame and change the values.
```
```suggestion
If you are unsure, copy the game.cfg from the testgame and change the values.
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+It will always contain all required and optional settings clearly marked.
+
+### 3. Make your game work
+* GameManager is always available in every script and has useful functions. (like `end_game` and `get_high_score()`)
```
```suggestion
* `GameManager` is always available in every script and has useful functions. (like `end_game` and `get_high_score()`)
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+
+### 3. Make your game work
+* GameManager is always available in every script and has useful functions. (like `end_game` and `get_high_score()`)
+* To save things that aren't scores while running the game, use `GameManager.get_game_data()` and modify the returned dictionary. This will get saved once end_game is called.
```
```suggestion
* To save things that aren't scores while running the game, use `GameManager.get_game_data()` and modify the returned dictionary. This will get saved once `end_game` is called.
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+* GameManager is always available in every script and has useful functions. (like `end_game` and `get_high_score()`)
+* To save things that aren't scores while running the game, use `GameManager.get_game_data()` and modify the returned dictionary. This will get saved once end_game is called.
+* Don't add Autoload scripts. They are loaded on startup and run all the time so it is bad practice to use one in a minigame.
+* The scene you defined in game.cfg will be loaded with get_tree().change_scene(). So It'll work like it is the mainscene.
```
```suggestion
* The scene you defined in game.cfg will be loaded using `get_tree().change_scene()`. So It'll work like it is the mainscene.
```

BlobCodes replied:
```suggestion
* The scene you defined in game.cfg will be loaded with get_tree().change_scene(). So it'll work like it is the mainscene.
```

BlobCodes replied:
```suggestion
* The scene you defined in game.cfg will be loaded with get_tree().change_scene(). So It'll work like it is the main scene.
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+* To save things that aren't scores while running the game, use `GameManager.get_game_data()` and modify the returned dictionary. This will get saved once end_game is called.
+* Don't add Autoload scripts. They are loaded on startup and run all the time so it is bad practice to use one in a minigame.
+* The scene you defined in game.cfg will be loaded with get_tree().change_scene(). So It'll work like it is the mainscene.
+* To return to the game selection use `GameManager.end_game("This is the end message", score)`, where score is a number that represents the score of the player.
```
```suggestion
* To return to the game selection, use `GameManager.end_game("This is the end message", score)`, where score is a number that represents the score of the player.
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+# How to add you own Minigame
```
```suggestion
# How to add your own minigame
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+# How to add you own Minigame
+## A Step by step Guide
```
You're using the headers incorrectly here.
Headers are meant to form a hierarchy (which is especially important for people with screen readers, etc).
Don't abuse the second header as a subtitle.

You could just merge the two lines together, for example: `Adding your own minigame - A step-by-step guide`.

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,19 @@
+### 3. Make your game work
+* GameManager is always available in every script and has useful functions. (like `end_game` and `get_high_score()`)
+* To save things that aren't scores while running the game, use `GameManager.get_game_data()` and modify the returned dictionary. This will get saved once end_game is called.
+* Don't add Autoload scripts. They are loaded on startup and run all the time so it is bad practice to use one in a minigame.
```
You're using `it is` here instead of `it's`.
Which variant you use is indifferent to me, but it should stay consistent (either only abbreviations like `it's`, `they're` and `isn't` or none).

BlobCodes replied:
```suggestion
* Don't add `AutoLoad` scripts. They are loaded on startup and run all the time so it is bad practice to use one in a minigame.
```

BlobCodes commented on game/games/testgame/balloonPop.gd:
```diff
@@ -3,7 +3,7 @@ extends MarginContainer
 const END_MESSAGE := "You got %s points!"
 const COLOR_MESSAGE := "Pop a %s balloon"
-const STATUS_MESSAGE := "You'r in stage %s/10 and have %s points!"
+const STATUS_MESSAGE := "You'r in stage %s/10 and have %s points! (Your highscore is: %s)"
```
```suggestion
const STATUS_MESSAGE := "You're in stage %s/10 and have %s points! (Your highscore is: %s)"
```

BlobCodes replied:
I just realized that this issue already exists in main.
Nevertheless, it's probably better to correct it.

BlobCodes reviewed (changes requested):
Spelling

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,18 @@
+# How to add your own minigame - A Step by step Guide
+
+## 1. Chose a folder
```
```suggestion
## 1. Choose a folder
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,18 @@
+# How to add your own minigame - A Step by step Guide
```
```suggestion
# How to add your own minigame - A step-by-step Guide
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,18 @@
+All files related only to your game go into a folder inside `res://games/`. You shouldn't rename it after the game is first merged. 
+There are no rules what to call it but a good idea is to use your name and the games name (ex. "asecondguy_assimilator") to avoid any accidental collisions.
+
+## 2. Make a game.cfg
```
```suggestion
## 2. Make a `game.cfg`
```

BlobCodes commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,18 @@
+There are no rules what to call it but a good idea is to use your name and the games name (ex. "asecondguy_assimilator") to avoid any accidental collisions.
+
+## 2. Make a game.cfg
+The game.cfg file resides in your minigame folder. It defines how your game is displayed. 
```
```suggestion
The `game.cfg` file resides in your minigame folder. It defines how your game is displayed. 
```

BlobCodes reviewed (changes requested):
Few more backticks

Numenter commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,169 @@ func _ready():
 
 
+func _save_data(file_name: String, data, game: String) -> int:
+	"""Save data to a file with for the given game.
```
I don't like the multiline comments, they look like strings. It's a preference i think.

Joshix-1 replied:
They are strings. I think they make more clear that they describe what the method does

Numenter replied:
I know it's for the latest godot version, they use ‘##’ for description.
https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_documentation_comments.html
But i think it”s a good guideline to use

Joshix-1 replied:
i didn't know that, thanks

Numenter replied:
I did not know this either :P

Numenter commented on game/menu/gamemanager.gd:
```diff
@@ -39,9 +39,8 @@ func load_game(game_cfg: ConfigFile):
-	"""Save data to a file with for the given game.
-	This is a private function, don't use this in a game
-	"""
+	## Save data to a file with for the given game.
```
On last thing and im happy.
Can you put the description on top of the function declaration?

Joshix-1 replied:
yes

MarioLiebisch commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,165 @@ func _ready():
+	if score != null:
+		if not "scores" in data:
+			data["scores"] = []
+		data["scores"].append([score, data["last_played"]])
```
The system should only track a limited amount of high scores (10? 20?) and store them sorted to avoid unnecessary processing and potentially moving around heaps of data over time or having to go through X entries to find the current high score.

Joshix-1 replied:
I agree with the sorting part, but as it isn't that much data, I think it would be best to save it all. Maybe we could in future display  graphs on how players did over time.

BjoernAkAManf replied:
I'd prefer being able to see the whole history part then though. Thinking of games like Tis-100, where some different stats are tracked globally and locally. Seeing how one progresses through time and space, would be quite cool. 

Might be out of scope for this project though.

Joshix-1 replied:
> I'd prefer being able to see the whole history part then though.

I agree, but I don't know how to do that properly, that can be done in the future. I don't want to put too much in this pr, and I'm not good with graphics stuff

MarioLiebisch commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,165 @@ func _ready():
 # return to the level select
-func end_game(message := "", _status = null):
+func end_game(message := "", score = null, _status = null):
+	var player_name = get_current_player()
```
`player_name` is defined, but never used.

MarioLiebisch commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,165 @@ func _ready():
+## Get time the current player played the currently running game.
+func get_played_time(game_id = null) -> float:
+	game_id = _last_loaded_game if game_id == null else game_id
+	assert(game_id != "")
```
This feels a bit odd. What's an unknown/unset `game_id`? Is it `null` or `""`? This should be consistent everywhere.

MarioLiebisch commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,165 @@ func _ready():
+
+	var scores = data["scores"]
+	var high_score = null
+	for score in scores:
```
See the other note, this list should be sorted, so for a player's personal high score, you should only have to return the very first entry rather than looking for it.

MarioLiebisch commented on game/menu/gamedisplay.gd:
```diff
@@ -9,10 +9,15 @@ func setup(game_cfg: ConfigFile):
 
 	var icon: Texture = load(
-		str(game_cfg.get_meta("folder_path"), game_cfg.get_value("game", "icon"))
+		str(
```
This feels really hard to read. It should only be a single format string, plus I'd use a fixed icon name for simplicity (being able to set it doesn't really add anything for the individual developer): `"res://games/%s/icon.png" % game_cfg.get_meta("folder_name")`

Joshix-1 replied:
I agree, but it's not the scope of this PR to change the configuration options

MarioLiebisch commented on game/menu/gamedisplay.gd:
```diff
@@ -34,5 +39,26 @@ func setup(game_cfg: ConfigFile):
 
+func update_text():
+	var game_id = game_file.get_meta("folder_name")
+	var text = (
```
This should also use one big text template to be filled, allowing easier localization, etc.

Joshix-1 replied:
One big template wouldn't allow for some of these things to be missing, if they aren't available

MarioLiebisch commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,165 @@ func _ready():
-
+	var scene = load(
+		(
+			"res://games/"
```
Again, should be a format string and I think it wouldn't hurt to just set a `main.tscn` as the default.

MarioLiebisch commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,165 @@ func _ready():
+	var cache_key = get_current_player() + "/" + game + "/" + file_name
+	var directory = "user://" + get_current_player() + "/" + game + "/"
+	Directory.new().make_dir_recursive(directory)
+	if cache_key in GAME_DATA_CACHE:
```
This check should happen before anything else here.

MarioLiebisch commented on game/menu/gamedisplay.gd:
```diff
@@ -9,10 +9,15 @@ func setup(game_cfg: ConfigFile):
 	game_file = game_cfg
 
 	$VBoxContainer/Label.text = game_cfg.get_value("game", "name")
```
Why isn't this line part of `update_text()`?

Joshix-1 replied:
this never changes

MarioLiebisch replied:
That's a point…🤔

MarioLiebisch commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,165 @@ func _ready():
+	var data = _load_data("game_meta_data", game_id)
+
+	if not "scores" in data:
+		return null
```
Shouldn't a non-existing high-score automatically be `0`?

Joshix-1 replied:
I don't think so, what if games allow for a negative score? I think there should be a difference

ASecondGuy replied:
Does it even have to be a number? Maybe games could define a custom score class?

Joshix-1 replied:
Currently I think it has to be a number. I don't see how a custom score class would be any benefit.

BjoernAkAManf replied:
I think one way or another one needs to define a way to compare a score to others. Commonly this can be achieved with conversion to a number. But what about a game like a way out, where the level / difficulty might influence your score as well?

A Class might allow easier implementation of that. However for classic mini games a high score is more common in arcade games, right?

Also prefer null instead of 0 for the initial score.

MarioLiebisch commented on game/games/testgame/README.md:
```diff
@@ -0,0 +1,18 @@
+All files related only to your game go into a folder inside `res://games/`. You shouldn't rename it after the game is first merged. 
+There are no rules what to call it but a good idea is to use your name and the games name (ex. "asecondguy_assimilator") to avoid any accidental collisions.
+
+## 2. Make a `game.cfg`
```
> Create a `game.cfg`

MarioLiebisch commented on game/games/testgame/balloonPop.gd:
```diff
@@ -3,7 +3,7 @@ extends MarginContainer
 const END_MESSAGE := "You got %s points!"
 const COLOR_MESSAGE := "Pop a %s balloon"
-const STATUS_MESSAGE := "You'r in stage %s/10 and have %s points!"
+const STATUS_MESSAGE := "You're in stage %s/10 and have %s points! (Your highscore is: %s)"
```
Numbers shouldn't be inserted as strings. Use `%d` instead.

MarioLiebisch commented on game/menu/gamemanager.gd:
```diff
@@ -1,7 +1,14 @@
-var _games = []
+## type of GAME_DATA_CACHE: Dictionary[String, Dictionary]
+##                                  (file name) (the data)
+const GAME_DATA_CACHE := {}  # NEVER MODIFY THIS, IF YOU DON'T KNOW WHAT YOU ARE DOING!
```
Not really sure about all these. `ALL_UPPERCASE` should ideally really be restricted to actual constant and never changing expressions.

Joshix-1 replied:
The linter doesn't like it when it's a const and not upper case. And I agree with that.

ASecondGuy commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,162 @@ func _ready():
+## This is a private function, don't use this in a game
+func _save_data(file_name: String, data, game: String) -> int:
+	var cache_key = get_current_player() + "/" + game + "/" + file_name
+	var directory = "user://" + get_current_player() + "/" + game + "/"
```
Maybe don't fill the user directory with the save folders.
Probably doesn't matter because I don't expect more than 5 accounts on the same machine.
```suggestion
	var directory = "user://savefiles/" + get_current_player() + "/" + game + "/"
```

ASecondGuy commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,162 @@ func _ready():
+		GAME_DATA_CACHE[cache_key] = _load_data(file_name, game)
+	GAME_DATA_CACHE[cache_key] = data
+	var file = File.new()
+	file.open(directory + file_name + ".json", File.WRITE)
```
```suggestion
	var err := file.open(directory + file_name + ".json", File.WRITE)
	if err != OK:
		return err
```
Don't expect writing to always work. There might be an ERR_PRINTER_ON_FIRE.

ASecondGuy commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,162 @@ func _ready():
+## This method is automatically called when a game ends.
+func save_game_data():
+	assert(_last_loaded_game != null)  # this should only be called if _last_loaded_game is set
+	_save_data("game_data", _load_data("game_data", _last_loaded_game), _last_loaded_game)
```
Isn't this already called whenever something is saved?
Maybe use a dirtybit instead to see what parts of the GAME_DATA_CACHE changed and save only those?

Joshix-1 replied:
This doesn't save the whole GAME_DATA_CACHE. This only saved the game_data for the current game.
The games can get a Dictionary which they can modify, the changes they do to this Dictionary will get saved

ASecondGuy commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +20,162 @@ func _ready():
 
+## Save data to a file with for the given game.
+## This is a private function, don't use this in a game
+func _save_data(file_name: String, data, game: String) -> int:
```
If it is the game_id please call it game_id and not game.
And you should write down somewhere that the game_id is the same as the game folders name.
```suggestion
func _save_data(file_name: String, data, game_id: String) -> int:
```

Joshix-1 replied:
Where would you write that down?

ASecondGuy replied:
Now that it is merged in `res://menu/README.md`

ASecondGuy reviewed (commented):
A few inconsistencies for you to fix.

Joshix-1 commented:
I've rebased the changes and squashed some commits to reduce the commit count to something less unreasonable.

BjoernAkAManf commented on game/menu/gamemanager.gd:
```diff
@@ -39,37 +190,36 @@ func _build_menu():
+	var err := config.load("res://games/" + folder_name + "/game.cfg")
 	if err != OK:
 		return err
-	f.set_meta("folder_path", path.get_base_dir() + "/")
```
We should keep the path though, so we can print the path in case of errors.

Joshix-1 replied:
The path would just be duplicate information

BjoernAkAManf replied:
Disagree, especially #37 needs it for visualization.

Joshix-1 replied:
You can just get the folder_path from the information in the meta.

ASecondGuy replied:
That PR isn't about visuals. So why would it need that for visualization?
Anyway, maybe we should include that path anyway just for convenience?
`cfg.get_meta("folder_path")` is easier than
`"res://games/%s/" % cfg.get_meta("folder_name")`.

Joshix-1 replied:
We could have a method that converts the folder name to the folder path. That would be nicer, with an optional file_name that can be added to the end

BjoernAkAManf replied:
I don't see how removing helpful information for error handling is a good idea.
I think we could live without explicit references to a config object and use getters and setters from our own class.
Then implementing folder_path could just return a templated string.
Otherwise removing that setting from the store seems counter productive.

Especially error messages a'la "Could not find res://games/meow/config.cfg are you sure it exists?" are better for developer experience than falling back to using debugger errors. Ignoring the obvious issue of warnings being issued for missing error handling.

Then again i could also live with a more capsuled approach to all of this, instead of relying on Strings inside a config object.

Joshix-1 replied:
> I don't see how removing helpful information for error handling is a good idea.

It is not removing any information. It just makes it a bit harder to access the information

BjoernAkAManf commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +22,164 @@ func _ready():
+
+	_save_data("game_meta_data", data, _last_loaded_game)
+
+	GAME_DISPLAYS[_last_loaded_game].update_text()
```
Might be a way to add multiple profiles (there was an PR right?) and keep the names here as well.

Joshix-1 replied:
I added a method that returns the current player (currently not changing), that can later be changed for multiple profiles

Numenter commented on game/games/testgame/balloonPop.gd:
```diff
@@ -1,9 +1,9 @@
 # warning-ignore-all:return_value_discarded
```
Don't know about this. Most return value warnings here are errors, better handle them.

Joshix-1 replied:
Why do you always correct stuff in my PR that was already there? I didn't change that 😅

Numenter replied:
good point

Numenter commented on game/menu/gamemanager.gd:
```diff
@@ -1,7 +1,16 @@
+##                                  (file name) (the data)
+const GAME_DATA_CACHE := {}  # NEVER MODIFY THIS, IF YOU DON'T KNOW WHAT YOU ARE DOING!
+const GAMES := {}  # type: Dictionary[String, ConfigFile]
+const GAME_DISPLAYS := {}  # type: Dictionary[String, gamedisplay]
```
Dictionarys and Arrays cant be "real" constants.
https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#constants
Make them private. No game should have access to other games data.
Make getter functions or give GameDisplays all needed data.

Numenter commented on game/menu/gamemanager.gd:
```diff
@@ -39,37 +190,36 @@ func _build_menu():
+		GAME_DISPLAYS[game_id] = display
 
 
 # go through every folder inside res://games/ and try to load the game.cfg into _games
 func _find_games():
```
I don't like this function name. It doesn't represent what the function does.
I suggest splitting the function.

- _find_games()
- _load_games()

ASecondGuy replied:
The Comment is outdated anyway. It isn't called _games for a few days now.
Still, I think it fits because a game is "found" when it is in the game list.
So finding all games requires writing them into the game list.

Joshix-1 replied:
I don't know how to split it, and again, this wasn't made by me. So I just renamed it to _find_and_load_games

Numenter commented on game/menu/gamemanager.gd:
```diff
@@ -39,37 +190,36 @@ func _build_menu():
 	var dir = Directory.new()
 	if dir.open("res://games") == OK:
 		dir.list_dir_begin(true)
 		var file_name = dir.get_next()
```
Why is this called file_name, when we're looking for folders?

ASecondGuy replied:
That is copy-pasted from the [Godot docs](https://docs.godotengine.org/en/stable/classes/class_directory.html#description).

Numenter commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +22,164 @@ func _ready():
 
+## Save data to a file with for the given game.
+## This is a private function, don't use this in a game
+func _save_data(file_name: String, data, game_id: String) -> int:
```
The whole data system, should probably be outsourced to another script

- onready var data_manager = preload(<DATA_MANAGER_PATH>).new()

Joshix-1 replied:
Moved most parts to another script

ASecondGuy commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +22,164 @@ func _ready():
 	_main.hide()
 
 
+## Save data to a file with for the given game.
```
```suggestion
## Save data to a file for the given game.
``` 
Fix Typo ;)

ASecondGuy commented on game/menu/gamemanager.gd:
```diff
@@ -13,22 +22,164 @@ func _ready():
+	var directory := SAVE_FILE_DIR % [get_current_player(), game_id]
+	Directory.new().make_dir_recursive(directory)
+	if not cache_key in GAME_DATA_CACHE:
+		GAME_DATA_CACHE[cache_key] = _load_data(file_name, game_id)
```
Why load this if you immediately overwrite it?

Numenter reviewed (commented):
Just some things that border me a little :P

Joshix-1 committed:
🏆️ add ability to store and display highscores


update highscore directly after getting a new one


🤦


💾 make functions for saving and loading game data more future proof


😅 rename meta stuff; don't put high_score in meta


🤦


display highscore in test game


allow running load_game_data and save_game_data without game argument


🤦


update some methods, to make it harder to screw up


Added "Add a minigame" guide

update documentation with new stuff


💎 make it really hard for games to destroy scores


🤦


change the way saving stuff works and add more meta data that gets saved


reset stuff sooner


make get_played_time return float instead of string


store game displays to only update the text when necessary


document some functions


¯\_(ツ)_/¯


game parameter → game_id


store data of games in their own files


two files per game to save everything + refactoring


Apply suggestions from code review

Co-authored-by: David Keller <davidkeller@tuta.io>
improve testgame readme further


other fixes


Update game/games/testgame/README.md

Co-authored-by: David Keller <davidkeller@tuta.io>
Apply suggestions from code review

Co-authored-by: David Keller <davidkeller@tuta.io>
Apply suggestions from code review

Co-authored-by: David Keller <davidkeller@tuta.io>
save files in better folder structure to make them easier to handle


🤦


use ## instead of multi-line string for method comments


move description to right place


follow some suggestions


use more string formatting


improve few things


fix typo | Apply suggestions from code review

Co-authored-by: ASecondGuy <asecondguy@gmail.com>
move big chunks of data managing code in own file


fix suggestion by ASecondGuy


improve way to get files of games


_find_games → _find_and_load_games


fix up some things
🤦


make it work again

ASecondGuy commented:
Why did you remove so many linebreaks in the READMEs?

Joshix-1 commented:
The diffs don't show missing line breaks, only missing trailing spaces

ASecondGuy commented:
Those trailing spaces are linebreaks in markdown.

ASecondGuy commented:
I also prefer `<br>` but last time I did that people told me to use double space.

Joshix-1 committed:
improvements after first rebase

improve documentation

re-add line breaks in READMEs


document game_id


more line breaks


improve highscore handling

Joshix-1 committed:
make saving meta data work with flappybird and pause menu

Numenter commented:
Not a fan of `<br>` in markdown files. HTML should be a fallback not the first choice.

ASecondGuy commented:
And maybe try to leave unrelated docu alone in a PR that isn't about updating docu.

Joshix-1 committed:
change html <br> to stupid invisible trailing spaces :)

Numenter commented:
This PR has 75% Votes (6/2) and is older than 3 Days. If its get rebased and the conflicts in flappy/player.gd solved. Then it should be merged.
@Joshix-1

Joshix-1 commented:
rebased and resolved conflicts :)
